demo-multiroot-react-18/src/MultiRootEditorDemo.tsx

```diff
@@ -0,0 +1,286 @@
+
+		for ( const [ rootName, { changedData, changedAttributes } ] of Object.entries( changedRoots ) ) {
+			if ( changedData ) {
+				newContent[ rootName ] = editor!.getData( { rootName } );
+			}
+
+			if ( changedAttributes && !hasNewAttributes ) {
+				hasNewAttributes = true;
+			}
+		}
+
+		setContent( { ...content, ...newContent } );
+
+		if ( hasNewAttributes ) {
+			setAttributes( { ...editor!.getRootsAttributes() } );
+		}
+	};
+
+	// Function to toggle read-only mode for selected root.
+	const toggleReadOnly = () => {
+		const root = editor!.model.document.selection.getFirstRange()!.root;
+
+		if ( !root || !root.rootName ) {
+			return;
+		}
+
+		const isReadOnly = disabledRoots.has( root.rootName );
+
+		if ( isReadOnly ) {
+			disabledRoots.delete( root.rootName );
+			editor!.enableRoot( root.rootName, SAMPLE_READ_ONLY_LOCK_ID );
+		} else {
+			disabledRoots.add( root.rootName );
+			editor!.disableRoot( root.rootName, SAMPLE_READ_ONLY_LOCK_ID );
+		}
+
+		setDisabledRoots( new Set( disabledRoots ) );
+	};
+
+	// Function to simulate an error in the editor.
+	// It is used to trigger watchdog to restart the editor.
+	const simulateError = () => {
+		setTimeout( () => {
+			const err: any = new Error( 'foo' );
+
+			err.context = editor;
+			err.is = () => true;
+
+			throw err;
+		} );
+	};
+
+	const addRoot = ( newRootAttributes: Record<string, unknown>, rootId?: string ) => {
+		const id = rootId || new Date().getTime();
+
+		for ( let i = 1; i <= numberOfRoots; i++ ) {
+			const rootName = `root-${ i }-${ id }`;
+
+			content[ rootName ] = '';
+			attributes[ rootName ] = { ...newRootAttributes, order: i * 10, row: id };
+		}
+
+		setContent( { ...content } );
+		setAttributes( { ...attributes } );
+		setNumberOfRoots( 1 );
+	};
+
+	const removeRoot = ( rootName: string ) => {
+		setContent( previousContent => {
+			const { [ rootName! ]: _, ...newContent } = previousContent;
+
+			return { ...newContent };
+		} );
+
+		setSelectedRoot( '' );
+	};
+
+	// Function to handle the addition of a new root element to the editor.
+	// This callback is necessary to correctly handle the multi root editor state.
+	const handleNewRoot = ( createRootElement: ( props?: Record<string, unknown> ) => JSX.Element ) => {
+		const element = createRootElement();
+
+		setElements( previousElements => [ ...previousElements, element ] );
+	};
+
+	// Function to handle the removal of a root element from the editor.
+	// This callback is necessary to correctly handle the multi root editor state.
+	const handleRemovedRoot = ( { rootName }: { rootName: string } ) => {
+		// Handling of undo operations.
+		if ( content[ rootName ] !== undefined ) {
+			removeRoot( rootName );
+		}
+
+		setElements( previousElements => previousElements.filter( element => element.props.id !== rootName ) );
+	};
+
+	// Grouping elements based on their row attribute and sorting them by order attribute.
+	// It is used only for presentation purposes.
```

Comment:
```suggestion
	// Grouping in a row is used for presentation purposes and you may remove it in actual integration.
	// However, we recommend ordering the roots, so that rows are put in a correct places when undo/redo is used.
	
```